crates/ruff_python_formatter/README.md

```diff
@@ -130,8 +141,67 @@ impl FormatNodeRule<ExprList> for FormatExprList {
 }
 ```
 
-Comments are categorized into `Leading`, `Trailing` and `Dangling`, you can override this in
-`place_comment`.
+A common challenge is that we want to attach comments to tokens (think keywords and syntactically
+meaningful characters such as `:`) that have no node on their own. A slightly simplified version of
+the `while` node in our AST looks like the following:
+
+```rust
+pub struct StmtWhile {
+    pub range: TextRange,
+    pub test: Box<Expr<TextRange>>,
+    pub body: Vec<Stmt<TextRange>>,
+    pub orelse: Vec<Stmt<TextRange>>,
+}
+```
+
+That means in
+
+```python
+while True:  # Trailing cond comment
+    if f():
+        break
+    # trailing while comment
+# leading else comment
+else:
+    print("while-else")
+```
+
+the `else` has no node, we're just getting the statements in its body.
+
+By default, the comment would get misattributed and moved to the `break` (the first node before the
+comment) or the `print` call (the first node after it). We avoid this by finding comments between
+two bodies that have the same indentation level as the keyword in
+`handle_in_between_bodies_own_line_comment` and marking them as dangling. Similarly, we find and
+mark comment after the colon(s). In `FormatStmtWhile`, we take the list of all dangling comments and
+split it into after-colon-comments, before-else-comments, etc. and manually insert them in the right
+position.
+
+A simplified implementation with only those two kinds comments:
+
+```rust
+fn fmt_fields(&self, item: &StmtWhile, f: &mut PyFormatter) -> FormatResult<()> {
+
+    // ...
+
+    let (trailing_condition_comments, or_else_comments) =
+        dangling_comments.split_at(or_else_comments_start);
```

Comment:
i think it would be helpful with at least some pseudocode/comment explaining how `or_else_comments_start` is calculated/constructed